pandas/core/indexes/base.py

```diff
@@ -1319,7 +1326,8 @@ def set_names(self, names, level=None, inplace=False):
 
         if level is not None and not is_list_like(level) and is_list_like(
                 names):
-            raise TypeError("Names must be a string")
+            msg = "Length of names must match number of levels in MultiIndex."
```

Comment:
I think the previous error message was more appropriate. May be better to say something like `Names must be a string when one level is provided`. But the new error message gives the impression that names should be a list, which is the not right.

Reply:
good. I will go for it